[WIP] Use ajax for loading large data

bourque opened the pull request:
This PR adds and changes several components of the web app to use asynchronous JavaScript and XML (ajax) to dynamically load data into the webpages.  It also adds a loading animation.  Currently a work in progress; more details to follow.

Closes #151

pep8speaks commented:
Hello @bourque, Thank you for submitting the Pull Request !

 - In the file [`jwql/website/apps/jwql/urls.py`](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py), following are the PEP8 issues :

> [Line 55:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L55): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 56:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L56): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 100 characters)
> [Line 57:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L57): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 58:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L58): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 100 characters)
> [Line 59:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L59): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)
> [Line 60:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L60): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (134 > 100 characters)
> [Line 62:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L62): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (129 > 100 characters)
> [Line 63:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L63): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (150 > 100 characters)
> [Line 64:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L64): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (138 > 100 characters)
> [Line 65:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L65): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (118 > 100 characters)
> [Line 66:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L66): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (133 > 100 characters)
> [Line 67:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L67): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 100 characters)
> [Line 68:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L68): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 100 characters)
> [Line 69:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L69): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)
> [Line 70:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/urls.py#L70): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (118 > 100 characters)

 - In the file [`jwql/website/apps/jwql/views.py`](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/views.py), following are the PEP8 issues :

> [Line 200:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/views.py#L200): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 100 characters)
> [Line 259:101](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/views.py#L259): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 100 characters)
> [Line 266:17](https://github.com/spacetelescope/jwql/blob/93fca554189ecd61381305528dbe11d1e7b89397/jwql/website/apps/jwql/views.py#L266): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

If you have not done so, please consult the [`jwql` Style Guide](https://github.com/spacetelescope/jwql/blob/master/style_guide/style_guide.md)

bourque commented:
Thanks @pep8speaks, glad to see you decided to start working again in 2019!

bourque commented:
@pep8speaks

bourque commented:
@pep8speaks suggest a diff

bourque commented:
I believe this PR is ready to be reviewed.  Changes include:

- Added new `get_base_url()` function to `utils.py` that returns a `base_url` based on the user running the code (i.e. if `admin_account`, return `dljwql.stsci.edu`, if not, return `localhost`
- Modified `test_api_views` to use new `get_base_url` function
- Rearranged functions in `utils.py` to be in alphabetical order
- Various minor edits based on `flake8` and @pep8speaks suggestions.
- Added `json_dumps_params` parameter to `JsonResponse()` calls in `api_views.py`, which makes the rendered JSON object be indented and nicely formatted
- Created a `thumbnails_ajax()` and `archive_thumbnails_ajax()` functions in `data_containers` and `views.py`, respectively to return data needed for the `ajax()` method in the `archive.html` and `thumbnails.html` templates
- Refactored `thumbnails()` function in `data_containers` to only return data that is otherwise not returned by `thumbnails_ajax()`
- Added a loading animation to the `archive.html` and `thumbnails.html` templates, along with supporting `css`.
- Added various javascript functions to `jwql.js`, `archive.html`, and `thumbnails.html` to support the loading of data dynamically using `ajax` method.
- Fixed indentation level in `show_only()` function in `jwql.js`
- Disabled the button to Unlooked images since that feature no longer works with these updates.

More on the last bullet: This PR currently breaks the Unlooked page, since the `ajax` method requires knowledge of which `proposal` is being used, and with unlooked images there is no way of knowing that.  I could spend some time and effort to build some workarounds, but I wonder if it is time to re-factor how we support the concept of "unlooked images".  We should probably better define what we mean by "unlooked", and how we want to keep track of which images have been looked at and which have not.  I'm inclined to tackle this issue in a future, separate PR.

I am assigning @laurenmarietta as reviewer since this is mostly web app related code.  @laurenmarietta I am aware that you will be away for some time, so no rush on this!  Apologies that this PR kind of snowballed into a big one, but most of these changes are related in some way!

laurenmarietta commented on jwql/website/apps/jwql/static/css/loader-magnify.css:
```diff
@@ -0,0 +1,74 @@
+  height: 200px !important;
+  -webkit-transform: translate(-100px, -100px) scale(1) translate(100px, 100px);
+  transform: translate(-100px, -100px) scale(1) translate(100px, 100px);
+}
```
Someday, I would love to hear how you did this magic!!

bourque replied:
@laurenmarietta This is pretty much copy pasta from the website that I got this from, so I would also like to know how this magic works! 👀

laurenmarietta commented on jwql/website/apps/jwql/templates/about.html:
```diff
@@ -57,6 +57,9 @@ <h4>Acknowledgements</h4>
 
+        <h4>Other attributions:</h4>
+        The loading animation 'magnify' is provided by <a href="https://loading.io">loading.io</a>
+
```
Unrelated thought: should we include an attribution to the Astropy project?

bourque replied:
@laurenmarietta Good idea!  I was thinking that we could also add an attribution for Ann Fields who helped us envision our logo improvements.  We should open a separate 'update-attributions' PR for these.

laurenmarietta commented on jwql/website/apps/jwql/templates/thumbnails.html:
```diff
@@ -2,6 +2,48 @@
+	 * Updates the thumbnail-array div with interactive images of thumbnails
+	 * @param {Object} data - The data returned by the update_thumbnails_page AJAX method
+	 */
+    function update_thumbnail_array(data) {
```
Same question here.

laurenmarietta commented on jwql/website/apps/jwql/static/js/jwql.js:
```diff
@@ -150,4 +157,93 @@ function sort_by_thumbnails(sort_type) {
+    content = 'Showing ' + num_rootnames + '/' + num_rootnames + ' activities';
+    content += '<a href="https://jwst-docs.stsci.edu/display/JDAT/File+Naming+Conventions+and+Data+Products" target="_blank" style="color: black">';
+    content += '<span class="help-tip mx-2">i</span></a>';
+    $("#img_show_count")[0].innerHTML = content;
```
I think I'm having a little trouble working out why, for the `thumbnails.html` page, divs for the `img_show_count` and sort options are written here in `jwql.js`, but similar/identical divs are still written in the HTML for `archive.html`. I would expect they would be handled in the same place? Sorry if I'm missing something simple!

bourque replied:
@laurenmarietta Good point.  Probably this javascript function could be generalized and used in both templates.  I'll try to do that.

laurenmarietta commented on jwql/website/apps/jwql/templates/archive.html:
```diff
@@ -2,73 +2,123 @@
-	</main>
+        <h3 id='title'>Archived {{ inst }} Images</h3>
+
+        <p id='img_show_count'></p>
```
Small thing: originally, the link and span below for the JDox tip were included in the `<p>` tag for the `img_show_count`, which kept the little (?) in line with the text about how many proposals were shown. I think the `</p>` here needs to be deleted for that to still work. Same with line 60 in `thumbnails.html`.

laurenmarietta commented on jwql/website/apps/jwql/static/js/jwql.js:
```diff
@@ -75,53 +81,53 @@ function search(n_proposals) {
 
```
In bad news, I am seeing that the search function isn't working correctly now. Typing in a particular number isn't accurately filtering out proposals that don't start with that number. Do you see this behavior, too? 

I'm not sure at the moment what might be causing it - maybe a misinterpretation of what's being typed into the search box, or a misassignment of the `proposal` attribute to each thumbnail div. If you don't have an immediate idea of why this might be happening, I'm happy to do a deeper dive.

laurenmarietta replied:
Though the sort functions are still working, interestingly...

bourque replied:
@laurenmarietta Hmmm, good catch, I'll look into this.

bourque commented on jwql/website/apps/jwql/templates/archive.html:
```diff
@@ -2,73 +2,123 @@
+     * @param {String} inst - The instrument of interest (e.g. "FGS")
+     * @param {String} base_url - The base URL for gathering data from the AJAX view.
+     */
+    function update_archive_page(inst, base_url) {
```
@laurenmarietta Ah yes, sorry I should have mentioned this before --

I am finding that any javascript function that references `src="{{static()}}"` in some way (e.g. `static("thumbnails/")` does not work when they live in `jwql.js`.  This seems to indicate that `static()` points to a path that is relative to the `website/apps/jwql/static/` directory instead of `website/apps/jwql/static/js/` directory.  Perhaps this has something to do with how we define the static paths in our `settings.py` file.  For now, I just kept things within the HTML template so that it works, but I do think we should figure out a better way in a separate PR, if you agree.

bourque commented:
@laurenmarietta My latest commit (82681a9) should address the issue with the JDox page help tip button and the suggestion to use a generalized `update_show_count` function in both `archive.html` and `thumbnails.html`.

The only lingering issue is that of the search bar not working as expected, though I don't seem to see that behavior.

laurenmarietta commented on jwql/website/apps/jwql/templates/archive.html:
```diff
@@ -2,73 +2,118 @@
+
+                    // Build div content
+                    content = '<div class="proposal text-center">';
+                    content += '<a href="/' + inst + '/archive/' + prop + '/" id="proposal' + i + '" proposal="' + prop + '"';
```
Okay, I figured out the search issue. Weird that still don't see it...

The javascript function that decides which thumbnails to hide and show when a value is entered into the search bar looks at the different id values for the thumbnails. These are `id=proposal#` where the `#` was originally taken from the loop iterator in Django, and started at 1. The new way (here) is a javascript for loop that starts at 0! 

When I tested it, changing the `i` in this line to `(i + 1)` fixed the problem.

bourque commented:
@laurenmarietta Ahhh, I understand now.  Sorry, before I was thinking you were referring to the search bar on the homepage.  I do see the issue on my end, and your suggested fix does resolve it.  I've fixed it in the latest commit (8ccbf2c).

bourque commented:
@laurenmarietta With your final approval, I think this is ready to merge.

laurenmarietta commented on jwql/website/apps/jwql/templates/archive.html:
```diff
@@ -2,73 +2,118 @@
+        <div class="d-flex flex-row">
+            <!-- Search Bar -->
+            <div class="input-group mb-3">
+              <input id='search_box' type="text" class="form-control" placeholder="Search" aria-label="Search by proposal" oninput="search('{{ num_proposals }}');">
```
So close!! Unfortunately, now I am seeing 
> Showing 17/{{ num_proposals }} proposals

when I type something into that search bar in `archive.html`. I think that it's because of how this `num_proposals` variable is getting passed into the javascript `search()` funtion.

Problem is, this looks completely correct to me. Do you have any ideas why that might be happening?

bourque replied:
@laurenmarietta Ahh good catch again! :scream: It is indeed because of the way `num_proposals` is being passed.  The problem is that the way I restructured things, the variable is no longer available in the view.  I'll have to come up with a workaround.

bourque commented:
@laurenmarietta Ok, turns out the fix was rather simple (4ec1016).
